### PR TITLE
Add buffer type

### PR DESCRIPTION
**Pull request by neithernut**
This PR adds an abstract buffer type, which serves as a unified interface for variaous buffers, both wayland based and internal ones.

### CODE REVIEW

File: src/compositor/buffer.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+void*
+ws_buffer_data(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return NULL;
+    }
+    if (!type->get_data) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: I guess this can be merged with the previous check...

File: src/compositor/buffer.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+int32_t
+ws_buffer_width(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return 0;
+    }
+    if (!type->get_width) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: This can be merged with the previous check, right?

File: src/compositor/buffer.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+int32_t
+ws_buffer_height(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return 0;
+    }
+    if (!type->get_height) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: Can also be merged with the previous check.

File: src/compositor/buffer.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+int32_t
+ws_buffer_stride(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return 0;
+    }
+    if (!type->get_stride) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: Same here again - merge with the previous check.

File: src/compositor/buffer.c [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+uint32_t
+ws_buffer_format(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return 0; //!< @todo: constant for invalid format
+    }
+    if (!type->get_format) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: Same here again - merge with the previous check.

File: src/compositor/buffer.c [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+void
+ws_buffer_begin_access(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return;
+    }
+    if (!type->begin_access) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: Same here again - merge with the previous check.

File: src/compositor/buffer.c [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+void
+ws_buffer_end_access(
+    struct ws_buffer* self
+) {
+    ws_buffer_type_id* type = (ws_buffer_type_id*) &self->obj.id;
+    if (!type) {
+        return;
+    }
+    if (!type->end_access) {
```
Comment (matthiasbeyer) [inline src/compositor/buffer.c]: Same here again - merge with the previous check.

File: src/compositor/buffer.h [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+ */
+struct ws_buffer {
+    struct ws_object obj;
+    struct ws_buffer_cbs* funcs;
+};
+
+/**
+ * Variable which holds the type information about the ws_wayland_obj type
+ */
+extern ws_buffer_type_id WS_OBJECT_BUFFER;
```
Comment (matthiasbeyer) [inline src/compositor/buffer.h]: The naming should be `WS_OBJECT_TYPE_ID_BUFFER` here, to fit the scheme.

File: src/compositor/buffer.h [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+ws_buffer_init(
+    struct ws_buffer* self //!< The buffer to initialize
+);
+
+/**
+ * Get a pointer to the buffer's contents
+ *
+ * @return pointer to the buffer's contents
+ *
+ * @warning to not pass NULL to this function! It will crash!
```
Comment (matthiasbeyer) [inline src/compositor/buffer.h:93]: Order of comment tags wrong, `@warning` first, `@return` as very last comment tag.

File: src/compositor/buffer.h [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Get the buffer's width
+ *
+ * @return width of the buffer's contents
+ *
+ * @warning to not pass NULL to this function! It will crash!
```
Comment (matthiasbeyer) [inline src/compositor/buffer.h:107]: Comment tag order, here too.

File: src/compositor/buffer.h [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Get the buffer's height
+ *
+ * @return height of the buffer's contents
+ *
+ * @warning to not pass NULL to this function! It will crash!
```
Comment (matthiasbeyer) [inline src/compositor/buffer.h:121]: Comment tag order, here too.

File: src/compositor/buffer.h [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Get the buffer's stride
+ *
+ * @return stride of the buffer's contents
+ *
+ * @warning to not pass NULL to this function! It will crash!
```
Comment (matthiasbeyer) [inline src/compositor/buffer.h:135]: Comment tag order, here too.

File: src/compositor/buffer.h [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+)
+__ws_nonnull__(1)
+;
+
+/**
+ * Get the buffer's format
+ *
+ * @return format of the buffer's contents
+ *
+ * @warning to not pass NULL to this function! It will crash!
```
Comment (matthiasbeyer) [inline src/compositor/buffer.h:149]: Comment tag order, here too.

**Comment by neithernut** [comment]
Do I have a GO for squashing?

**Comment by matthiasbeyer** [comment]
Yes.

**Comment by neithernut** [comment]
Merging so I can progress further.

**Pull request merged**